batch/batch/driver/main.py

```diff
@@ -1322,6 +1330,139 @@ async def monitor_system(app):
     monitor_instances(app)
 
 
+async def compact_agg_billing_project_users_table(db: Database):
+    async def compact(key: Optional[Tuple[str, str, int]]) -> Optional[Tuple[str, str, int]]:
+        if key:
+            billing_project, user, resource_id = key
+            where_condition = (
+                'AND ((billing_project > %s) OR'
+                '(billing_project = %s AND `user` > %s) OR'
+                '(billing_project = %s AND `user` = %s AND resource_id > %s))'
+            )
+            args = [billing_project, billing_project, user, billing_project, user, resource_id]
+        else:
+            where_condition = ''
+            args = []
+
+        token = secret_alnum_string(5)
+        scratch_table_name = f'`scratch_{token}`'
+
+        try:
+            next_key = await db.execute_and_fetchone(
+                f'''
+CREATE TEMPORARY TABLE {scratch_table_name} ENGINE=MEMORY
+AS (SELECT billing_project, `user`, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+    FROM aggregated_billing_project_user_resources_v3
+    WHERE token != 0 {where_condition}
```

Comment:
This approach is conceptually very clear but has a bit of a stop-the-world effect for a particular batch. Maybe it's not a big deal so I'm happy to start with this solution, but if the blocking ever creates noticeable halts in throughput we might want to start randomly sampling subsets of the tokens so we don't lock all the rows for a particular resource at one point in time.